libdnf/rpm/solv_query.cpp

```diff
@@ -986,6 +993,357 @@ SolvQuery & SolvQuery::ifilter_location(libdnf::sack::QueryCmp cmp_type, std::ve
+        // Removal of NOT CmpType makes following comparissons easier and effective
+        cmp_type = cmp_type - libdnf::sack::QueryCmp::NOT;
+    }
+
+    ReldepList reldep_list(p_impl->sack.get());
+
+    str2reldep_internal(reldep_list, cmp_type, patterns);
+
+    if (cmp_not) {
+        return ifilter_provides(libdnf::sack::QueryCmp::NEQ, reldep_list);
+    } else {
+        return ifilter_provides(libdnf::sack::QueryCmp::EQ, reldep_list);
+    }
+}
+
+void SolvQuery::Impl::filter_reldep(
+    Id libsolv_key, libdnf::sack::QueryCmp cmp_type, const std::vector<std::string> & patterns) {
+    bool cmp_not = (cmp_type & libdnf::sack::QueryCmp::NOT) == libdnf::sack::QueryCmp::NOT;
+    if (cmp_not) {
+        // Removal of NOT CmpType makes following comparissons easier and effective
+        cmp_type = cmp_type - libdnf::sack::QueryCmp::NOT;
+    }
+
+    ReldepList reldep_list(sack.get());
+    str2reldep_internal(reldep_list, cmp_type, patterns);
+    if (cmp_not) {
+        filter_reldep(libsolv_key, libdnf::sack::QueryCmp::NEQ, reldep_list);
+    } else {
+        filter_reldep(libsolv_key, libdnf::sack::QueryCmp::EQ, reldep_list);
+    }
+}
+
+void SolvQuery::Impl::filter_reldep(Id libsolv_key, libdnf::sack::QueryCmp cmp_type, const ReldepList & reldep_list) {
+    bool cmp_not = false;
+    switch (cmp_type) {
+        case libdnf::sack::QueryCmp::EQ:
+            break;
+        case libdnf::sack::QueryCmp::NEQ:
+            cmp_not = true;
+            break;
+
+        default:
+            throw SolvQuery::NotSupportedCmpType("Used unsupported CmpType");
+    }
+
+    solv::SolvMap filter_result(static_cast<int>(sack->pImpl->get_nsolvables()));
+    Pool * pool = sack->pImpl->get_pool();
+
+    sack->pImpl->make_provides_ready();
+
+    solv::IdQueue rco;
+
+    for (PackageId candidate_id : query_result) {
+        Solvable * solvable = solv::get_solvable(pool, candidate_id);
+        auto reldep_list_size = reldep_list.size();
+        for (int index = 0; index < reldep_list_size; ++index) {
+            Id reldep_filter_id = reldep_list.get_id(index).id;
+
+            rco.clear();
+            solvable_lookup_idarray(solvable, libsolv_key, &rco.get_queue());
+            auto rco_size = rco.size();
+            for (int index_j = 0; index_j < rco_size; ++index_j) {
+                Id reldep_id_from_solvable = rco[index_j];
+
+                if (pool_match_dep(pool, reldep_filter_id, reldep_id_from_solvable) != 0) {
+                    filter_result.add_unsafe(candidate_id);
+                    break;
+                }
+            }
+        }
+    }
+
+    // Apply filter results to query
+    if (cmp_not) {
+        query_result -= filter_result;
+    } else {
+        query_result &= filter_result;
+    }
+}
+
+void SolvQuery::Impl::filter_reldep(Id libsolv_key, libdnf::sack::QueryCmp cmp_type, const PackageSet & package_set) {
+    bool cmp_not = false;
+    switch (cmp_type) {
+        case libdnf::sack::QueryCmp::EQ:
```

Comment:
Only small speed optimization. Set value `cmp_not = false` at this place and remove value from declaration above `bool cmp_not`.

Reply:
Fixed